pipenv/project.py

```diff
@@ -267,7 +267,18 @@ def virtualenv_exists(self):
     def get_location_for_virtualenv(self):
         if self.is_venv_in_project():
             return os.path.join(self.project_directory, ".venv")
-        return str(get_workon_home().joinpath(self.virtualenv_name))
+
+        name = self.virtualenv_name
+        if self.project_directory:
+            venv_path = os.path.join(self.project_directory, ".venv")
+            if os.path.isfile(venv_path):
```

Comment:
If you use `Path(venv_path)`  you can use the check `.exists() and not Path(venv_path).isdir()` for files. You should do the same path object instantiation on the files contents to make sure the virtualenv exists.

Reply:
Why `not isdir()` rather than just `isfile()`? Furthermore, how would using pathlib over os.path help? Seems to me like an equivalent check can be written with os.path. I like the pathlib but I've used os.path just to remain consistent with the rest of the code.

> You should do the same path object instantiation on the files contents to make sure the virtualenv exists.

Is this just because the path wont be created if it doesn't exist when the venv is being created?

Reply:
There are edge cases where `isfile` and `isdir` can both be false, such as [device files](https://en.wikipedia.org/wiki/Device_file). While it doesn’t really matter practically, it is semantically more correct to check for “not directory” because those special files are still readable. It also matches the directory check in `is_venv_in_project`, which makes the code easier to reason with.

I’m wondering though, it may be a better idea to merge the two checks. `is_venv_in_project` already checks for directory, so when you reach this part, an `exists` check is enough. We shouldn’t reach here if it’s a directory; if it exists, we want to try reading it (and fails with a straight-up error if that fails).

Reply:
I agree that it's better to rely on `is_venv_in_project`'s directory check. `exists` should be sufficient here.

What should be done when the directory specified in a .venv file does not exist? That hasn't been addressed yet.

Reply:
It would need to fail back to use `self.virtualenv_name`. Now you mentioned it, I realise the code is a bit tangled here and the implementation doesn’t work well. I’ll try to find some time to sort the structure out so this can progress.

Reply:
`virtualenv` just creates the directory if it doesn't exist. I consider that to be a feature, so I don't think we need to fall back to `self.virtualenv_name` after all.